src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -681,6 +681,178 @@ private bool TryInsert(TKey key, TValue value, InsertionBehavior behavior)
             return true;
         }
 
+        internal ref TValue? GetValueRefOrAddDefault(TKey key, out bool exists)
```

Comment:
Ideally, this would not add any new code to Dictionary to make this niche method pay-for-play.

Reply:
I was hoping for `Dictionary` to already have an equivalent method to just expose like it was done for `GetValueRefOrAddRef`, but unfortunately there doesn't seem to be one and I'm not really seeing a way to implement this API without either adding some code (which I did so far) or at least making some changes to `TryInsert` (which is the most similar one) and using that. The reason I didn't do that though was that that would've likely led to some micro-benchmark regressions in other scenarios where the additional logic to return a ref wasn't needed. Not really sure how to proceed here, feedbacks are more than welcome 😄

One random idea I had to move the logic out of the `Dictionary` type was to refactor the code to be like:
```csharp
class Dictionary
{
    internal class Impl
    {
        public internal ref TValue? GetValueRefOrAddDefault(TKey key, out bool exists);
    }
}
```

So that `Impl.GetValueRefOrAddDefault` could access private members of the dictionary but without the method adding overhead to the dictionary type itself. Though this approach is adding another type, so it might not be desireable either. Another idea was to move the whole method logic to `CollectionsMarshal`, but that'd have required to make all the necessary dictionary members internal (and not private), so that seemed at the very least quite a bit ugly 😕

Thoughts?

Reply:
> without either adding some code (which I did so far) or at least making some changes to TryInsert (which is the most similar one) and using that. The reason I didn't do that though was that that would've likely led to some micro-benchmark regressions in other scenarios where the additional logic to return a ref wasn't needed.

Now you know why @benaadams didn't add this ;-)

Reply:
It was a trap 🤣